Docs - Azure join method (#21597)
gravitational/teleport: docs/pages/management/guides/joining-nodes-azure.mdx

---
title: Joining Nodes via Azure Managed Identity
description: Use the Azure join method to add Nodes to your Teleport cluster on Azure
---

This guide will explain how to use the **Azure join method** to configure
Teleport Nodes and Proxy Service instances to join your Teleport cluster without
sharing any secrets when they are running in an Azure Virtual Machine.

The Azure join method is available in Teleport 12.1+. It is available to any
Teleport Node or Proxy running in an Azure Virtual Machine. The Virtual Machine
must have a
[Managed Identity](https://learn.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/overview)

> assigned to it with permission to read virtual machine info.

Reviewer: > The Virtual Machine
must have a
[Managed Identity](https://learn.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/overview)
assigned to it with permission to read virtual machine info.

I think we should move this sentence to the Prerequisites section so all prerequisites are listed in one place.

<Details
scope={["oss", "enterprise", "cloud"]}
scopeOnly
opened
title="Another Azure Node joining method"
>

You can also configure Nodes running in Azure to join a cluster via **secret tokens**,
which is useful when you don't want to rely on Azure-specific APIs.
Read more in the following guide:

> [Adding Nodes to the cluster](../admin/adding-nodes.mdx)

Reviewer: ```suggestion
[Adding Nodes to your Cluster](../admin/adding-nodes.mdx).
```

</Details>

<Notice type="warning" scope={["oss", "enterprise"]}>

The Azure join method will not work if TLS is terminated at a load balancer in
front of your Teleport Proxy Service unless the Node using this method is
connecting directly to the Auth Service.

</Notice>

## Prerequisites

(!docs/pages/includes/edition-prereqs-tabs.mdx!)

- An Azure Virtual Machine to act as a Teleport Node, with the Teleport binary
  installed.

(!docs/pages/includes/tctl.mdx!)

> ## Step 1/4. Set up Managed Identity.

Reviewer: ```suggestion
## Step 1/4. Set up a Managed Identity
```

Every virtual machine hosting a Node or Proxy using the Azure method to join
your Teleport cluster needs a Managed Identity assigned to it. The identity 

> requires at least the `Microsoft.Compute/virtualMachines/read` permission.

Reviewer: > The identity 
requires at least the...

Do we have any guidance on when a user should add a permission besides `Microsoft.Compute/virtualMachines/read` to the Managed Identity?

I think it would also make this paragraph a bit clearer to indicate how the Node/Proxy uses the `Microsoft.Compute/virtualMachines/read` permission.

> To create a managed identity with an appropriate role, first

Reviewer: Other Azure-related guides in the docs include these instructions within the page, which I think makes it easier for a user to work linearly through the steps. An example is here: https://goteleport.com/docs/application-access/cloud-apis/azure/#step-14-grant-an-identity-to-your-vm

[create a custom role](https://learn.microsoft.com/en-us/azure/role-based-access-control/custom-roles)
with the `Microsoft.Compute/virtualMachines/read` permission.
Then [create a new user-assigned managed identity](https://learn.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/how-manage-user-assigned-managed-identities?pivots=identity-mi-methods-azp#create-a-user-assigned-managed-identity)
and [assign it your custom role](https://learn.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/how-manage-user-assigned-managed-identities?pivots=identity-mi-methods-azp#manage-access-to-user-assigned-managed-identities).
Note the client ID of the identity for Step 3. Finally,
[assign the identity to the VM that will join your Teleport cluster](https://learn.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/qs-configure-portal-windows-vm).

<Admonition
  type="warning"
  title="IMPORTANT"
>

> As part of authentication, Teleport will be able to do anything allowed by the

Reviewer: > As part of authentication, Teleport will be able to do anything allowed by the identity. 

Does this mean that, when Teleport authenticates to Azure, it will be able to list all virtual machines (using the `read` permissions described above)?

Author: Yes.

Reviewer: Do we have any instructions we can provide for limiting the managed identity's permissions to only what is necessary? We already instruct the reader to scope the managed identity to a resource group/region, for example.

Author: Limiting the scope and permissions should be enough. This warning is more geared towards readers using an identity other than the one created in the guide.

Reviewer: Makes sense, thanks!

identity. We strongly recommend limiting the managed identity's permissions
to only what is necessary. 
</Admonition>

## Step 2/4. Create the Azure Node joining token
Under the hood, Nodes will prove that they are running in your Azure subscription
by sending a signed attested data document and access token to the Teleport Auth

> server. The Node's identity must match an allow rule configured in your Azure

Reviewer: ```suggestion
Service. The Node's identity must match an allow rule configured in your Azure
```

Node joining token.

Create the following `token.yaml` with an `allow` rule specifying your Azure
subscription and the resource group that your Node's identity must match.

```yaml
# token.yaml
kind: token
version: v2
metadata:
  # the token name is not a secret because instances must prove that they are
  # running in your Azure subscription to use this token
  name: azure-token
  # set a long expiry time, the default for tokens is only 30 minutes
  expires: "3000-01-01T00:00:00Z"
spec:
  # use the minimal set of roles required
  roles: [Node]
  azure:
    allow:
      # specify the Azure subscription which Nodes may join from
      - subscription: 11111111-1111-1111-1111-111111111111
      # multiple allow rules are supported
      - subscription: 22222222-2222-2222-2222-222222222222

>       # resource_groups is optional and allows you to restrict the resource group of joining Nodes

Reviewer: ```suggestion
      # resource_groups is optional and allows you to restrict the resource group of
      # joining Nodes
```
So this comment doesn't wrap

      - subscription: 33333333-3333-3333-3333-333333333333
        resource_groups: ["group1", "group2"]
```

The token name `azure-token` is just an example and can be any value you want to
use, as long as you use the same value for `join_params.token_name` in Step 3.

> Run `tctl create -f token.yaml` to create the token.

Reviewer: ````suggestion
Run the following command  to create the token:

```code
$ tctl create -f token.yaml
```
````

So a user can copy/paste this more easily

---
title: Joining Nodes via Azure Managed Identity
description: Use the Azure join method to add Nodes to your Teleport cluster on Azure
---

This guide will explain how to use the **Azure join method** to configure
Teleport Nodes and Proxy Service instances to join your Teleport cluster without
sharing any secrets when they are running in an Azure Virtual Machine.

The Azure join method is available in Teleport 12.1+. It is available to any
Teleport Node or Proxy running in an Azure Virtual Machine.

<Details
scope={["oss", "enterprise", "cloud"]}
scopeOnly

> opened

Reviewer 2: Let's change the title here. Token based joining isn't an "Azure joining method" - it's a generic join method that works on all infra.

title="Another Azure Node joining method"
>

You can also configure Nodes running in Azure to join a cluster via **secret tokens**,
which is useful when you don't want to rely on Azure-specific APIs.
Read more in the following guide:
[Adding Nodes to your Cluster](../admin/adding-nodes.mdx)

</Details>

<Notice type="warning" scope={["oss", "enterprise"]}>

The Azure join method will not work if TLS is terminated at a load balancer in
front of your Teleport Proxy Service unless the Node using this method is
connecting directly to the Auth Service.

</Notice>

## Prerequisites

(!docs/pages/includes/edition-prereqs-tabs.mdx!)

- An Azure Virtual Machine to act as a Teleport Node, with the Teleport binary
  installed. The Virtual Machine must have a
  [Managed Identity](https://learn.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/overview)
  assigned to it with permission to read virtual machine info.

(!docs/pages/includes/tctl.mdx!)

> ## Step 1/4. Set up a Managed Identity.

Reviewer: ```suggestion
## Step 1/4. Set up a Managed Identity
```
We don't include punctuation at the end of docs headings

Every virtual machine hosting a Node or Proxy using the Azure method to join
your Teleport cluster needs a Managed Identity assigned to it. The identity 

> requires at least the `Microsoft.Compute/virtualMachines/read` permission so

Reviewer: ```suggestion
requires the `Microsoft.Compute/virtualMachines/read` permission so
```
Since we say that "No other permissions are required".

can look up the virtual machine. No other permissions are required.

### Create a custom role

Enter the name of your Azure resource group in the Azure Portal search box and
visit the page for that resource group. On the left navigation sidebar, click
the **Access control (IAM)** tab. In the row of buttons at the top of the
**Access control (IAM)** panel, click **Add > Add custom role**.

In the **Custom role name** field, enter `teleport-read-vm`.

![Create custom
role](../../../img/azure/create-custom-role.png)

Click the **Permissions** tab, then within the **Permissions** view, click
**+Add permissions**.

Enter `Microsoft.Compute/virtualMachines/read` in the search box. Click the
**Microsoft Compute** box, then enable **Read: Get Virtual Machine**. Click
**Add**.

![Add virtual machine read
permission](../../../img/azure/read-permission.png)

Click **Review + create**, then **Create**.

### Create an Azure managed identity

Visit the [Managed
Identities](https://portal.azure.com/#view/HubsExtension/BrowseResource/resourceType/Microsoft.ManagedIdentity%2FuserAssignedIdentities)
view in Azure Portal.

> Click **Create**.

Reviewer 2: These manual instructions are fine, but don't forget about infra as code. A terraform or ARM snippet for creating these resources would be helpful.

the new identity listed in the **User assigned** sub-tab:

![Verifying that you added the
identity](../../../img/application-access/azure/verify-id.png)

<Admonition
  type="warning"
  title="IMPORTANT"
>
As part of authentication, Teleport will be able to do anything allowed by the
identity. If you use a managed identity other than the one created in this
guide, we strongly recommend limiting its permissions and scope. 
</Admonition>

## Step 2/4. Create the Azure Node joining token
Under the hood, Nodes will prove that they are running in your Azure subscription
by sending a signed attested data document and access token to the Teleport Auth
Service. The Node's identity must match an allow rule configured in your Azure
Node joining token.

Create the following `token.yaml` with an `allow` rule specifying your Azure
subscription and the resource group that your Node's identity must match.

```yaml
# token.yaml
kind: token
version: v2
metadata:
  # the token name is not a secret because instances must prove that they are
  # running in your Azure subscription to use this token
  name: azure-token
  # set a long expiry time, the default for tokens is only 30 minutes

>   expires: "3000-01-01T00:00:00Z"

Reviewer 2: Is this necessary? I think @strideynet may have made some changes to make it optional.

spec:
  # use the minimal set of roles required
  roles: [Node]
  azure:
    allow:
      # specify the Azure subscription which Nodes may join from
      - subscription: 11111111-1111-1111-1111-111111111111
      # multiple allow rules are supported
      - subscription: 22222222-2222-2222-2222-222222222222
      # resource_groups is optional and allows you to restrict the resource group of
      # joining Nodes
      - subscription: 33333333-3333-3333-3333-333333333333
        resource_groups: ["group1", "group2"]
```

The token name `azure-token` is just an example and can be any value you want to
use, as long as you use the same value for `join_params.token_name` in Step 3.

Run the following command to create the token:

```code
$ tctl create -f token.yaml
```

## Step 3/4. Configure your Nodes
The Azure join method can be used for Teleport services running the SSH, Proxy,

> Kubernetes, Application, or Database Service.

Reviewer 2: Why not desktop?

---
title: Joining Nodes via Azure Managed Identity
description: Use the Azure join method to add Nodes to your Teleport cluster on Azure
---

This guide will explain how to use the **Azure join method** to configure
Teleport Nodes and Proxy Service instances to join your Teleport cluster without
sharing any secrets when they are running in an Azure Virtual Machine.

> The Azure join method is available in Teleport 12.1+. It is available to any

Reviewer 2: Does it work for other services too? (DB, desktop, bot)

        "permissions": [
          {
            "actions": [
              "Microsoft.Compute/virtualMachines/read"
            ],
            "notActions": []
          }
        ],
        "assignableScopes": [
          "[resourceGroup().id]"
        ]
      }
    },
    {
      "type": "Microsoft.ManagedIdentity/userAssignedIdentities",
      "name": "[parameters('identityName')]",
      "apiVersion": "2018-11-30",
      "location": "[resourceGroup().location]"
    }
  ],
  "outputs": {
    "principalID": {
      "type": "string",
      "value": "[reference(resourceId('Microsoft.ManagedIdentity/userAssignedIdentities', parameters('identityName'))).principalId]"
    },
    "clientID": {
      "type": "string",
      "value": "[reference(resourceId('Microsoft.ManagedIdentity/userAssignedIdentities', parameters('identityName'))).clientId]"
    },
    "roleName": {
      "type": "string",
      "value": "[variables('roleDefName')]"
    }
  }
}
```

Then run the following commands to create the custom role and managed identity:

```code
$ DEPLOY_OUTPUT=$(az deployment group create \
--resource-group <your-resource-group> \
--template-file teleport-create-identity.json)

> $ PRINCIPAL_ID=$(echo $DEPLOY_OUTPUT | jq -r '.properites.outputs.principalID.value')

Reviewer 3: I'd add that `jq` needs to be installed. Otherwise, some people are not familiar with it may end up with `PRINCIPAL_ID=jq not found`

      }
    }
  ]
}
```

Run the following command to assign your custom role to your managed identity
and assign the identity to your virtual machines:

```code
$ az deployment group create \
--resource-group <your-resource-group> \
--template-file teleport-assign-identity.json \
-- parameters principalId="$PRINCIPAL_ID" roleName="$ROLE_NAME" \
vmNames='<list-of-vm-names>'
```

Use the value of `CLIENT_ID` in Step 3.
</TabItem>
</Tabs>

## Step 2/4. Create the Azure Node joining token
Under the hood, Nodes will prove that they are running in your Azure subscription
by sending a signed attested data document and access token to the Teleport Auth
Service. The Node's identity must match an allow rule configured in your Azure
Node joining token.

Create the following `token.yaml` with an `allow` rule specifying your Azure
subscription and the resource group that your Node's identity must match.

```yaml
# token.yaml
kind: token
version: v2
metadata:
  # the token name is not a secret because instances must prove that they are
  # running in your Azure subscription to use this token
  name: azure-token
  # set a long expiry time, the default for tokens is only 30 minutes

>   expires: "3000-01-01T00:00:00Z"

Reviewer 2: @strideynet this is no longer necessary, right?

Reviewer 3: @strideynet Should we remove it from master and only added when backporting to v12? Sounds like adding it everywhere just encourages "bad practices" that are not longer necessary.

Reviewer 4: That sounds like a good plan !

      - subscription: 22222222-2222-2222-2222-222222222222
      # resource_groups is optional and allows you to restrict the resource group of
      # joining Nodes
      - subscription: 33333333-3333-3333-3333-333333333333
        resource_groups: ["group1", "group2"]
```

The token name `azure-token` is just an example and can be any value you want to
use, as long as you use the same value for `join_params.token_name` in Step 3.

Run the following command to create the token:

```code
$ tctl create -f token.yaml
```

## Step 3/4. Configure your Nodes
The Azure join method can be used for Teleport services running the SSH, Proxy,
Kubernetes, Application, Database, or Desktop Service.

Configure your Teleport Node with a custom `teleport.yaml` file. Use the
`join_params` section with `token_name` matching your token created in Step 2
and `method: azure` as shown in the following example config:

```yaml
# /etc/teleport.yaml
version: v3
teleport:
  join_params:
    token_name: azure-token
    method: azure
    azure:
      # client_id is the client ID of the managed identity created in Step 1.
      client_id: 11111111-1111-1111-1111-111111111111

>   proxy_server: https://teleport.example.com:443

Reviewer 2: I don't think this should have the https scheme on it.

  title="IMPORTANT"
>
As part of authentication, Teleport will be able to do anything allowed by the
identity. If you use a managed identity other than the one created in this

> guide, we strongly recommend limiting the its permissions and scope. 

Reviewer: ```suggestion
guide, we strongly recommend limiting its permissions and scope. 
```
